### PR TITLE
[new feature] logql engine support exec `vector(0)` grama

### LINKED ISSUE #6946: Implement vector() function
**Is your feature request related to a problem? Please describe.**
I want to turn null (no data) to zero in query result in easy and reliable way.

There are other Loki users who want the same.
https://community.grafana.com/t/how-can-i-turn-no-data-to-zero-in-loki/40694

**Describe the solution you'd like**
Implement a function like [vector()](https://prometheus.io/docs/prometheus/latest/querying/functions/#vector) in PromQL.
> vector(s scalar) returns the scalar s as a vector with no labels.
Then users can join their query with vector() to turn NULL to zero, as below.
```
vector1
or
vector(0)
```

**Describe alternatives you've considered**
(1) Grafana's transform feature.
However, this feature only works for a dashboard panel with single Loki query. It doesn't work for multiple queries for a panel. 

(2) Tried to generate zeros reliably, but not perfect
```
sum(count_over_time({namespace="traefik"}|="foo"[1m])) 
or
sum(count_over_time({namespace="traefik"}[1m])) < bool 0
```
The 2nd vector can still return NULL if there is no log data matching labels sets. Not useful for log streams which has fluctuating log ingestion. 

(3) Tried absent_over_time() as 2nd vector instead, it didn't work for me.
- Bug report: [Using absent_over_time with logical operator returns "Unimplemented" error. #6943](https://github.com/grafana/loki/issues/6943)

(4) First, stored the log-metrics in Prometheus. Second, use vector() in Prometheus. 
This works of course. But I wish I can do the same using LokiQL which is upstream.

**Additional context**

### PR DESCRIPTION
**Pull request by liguozhong**
**What this PR does / why we need it**:
logql engine support exec `vector(0)` grama.
new PR of :https://github.com/grafana/loki/pull/7007
**Which issue(s) this PR fixes**:
Fixes #6946

**Special notes for your reviewer**:
test passed.
test1: vectorExpr only
image

test2: binExpr and vectorExpr
image

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: pkg/logql/evaluator.go [hunk 1]
```diff
@@ -207,6 +207,12 @@ func (ev *DefaultEvaluator) StepEvaluator(
 		return binOpStepEvaluator(ctx, nextEv, e, q)
 	case *syntax.LabelReplaceExpr:
 		return labelReplaceEvaluator(ctx, nextEv, e, q)
+	case *syntax.VectorExpr:
+		val, err := e.Value()
+		if err != nil {
+			return nil, err
```
Comment (dannykopping) [inline pkg/logql/evaluator.go:213]: Can you please add a test-case where an unexpected value is used, to cover this case?

**Review by dannykopping — COMMENTED** [review]
Looks great @liguozhong, one comment

File: pkg/logql/evaluator.go (hunk 1 above)
Comment (liguozhong) [reply pkg/logql/evaluator.go:213]: ok,done.

Comment (liguozhong) [reply pkg/logql/evaluator.go:213]: ok,done

Comment (dannykopping) [reply pkg/logql/evaluator.go:213]: Thanks a lot!

**Review by dannykopping — COMMENTED** [review]
LGTM, thanks for the great feature @liguozhong :tada:
Can you please add a CHANGELOG entry, and some documentation?
Or do you want to do that in a separate PR?

**Comment by liguozhong** [comment]
> LGTM, thanks for the great feature @liguozhong 🎉 Can you please add a CHANGELOG entry, and some documentation? Or do you want to do that in a separate PR?

I want to complete the documentation work and CHANGELOG in another PR. and the PR of the documentation will need @KMiller-Grafana 's help, I am not good enough to independently complete an easy-to-understand documentation to explain how vector(0) applies to oncall's in the police.

**Review by owen-d — APPROVED** [review]
Nice, LGTM

**Comment by dannykopping** [comment]
> > LGTM, thanks for the great feature @liguozhong tada Can you please add a CHANGELOG entry, and some documentation? Or do you want to do that in a separate PR?
> 
> I want to complete the documentation work and CHANGELOG in another PR. and the PR of the documentation will need @KMiller-Grafana 's help, I am not good enough to independently complete an easy-to-understand documentation to explain how vector(0) applies to oncall's in the police.

OK, thank you. @KMiller-Grafana is no longer with the company, but I will help you with the documentation.

**Pull request merged**